Sources/AuthFoundation/Security/Internal/Keychain+Extensions.swift

```diff
@@ -14,13 +14,158 @@
+
+protocol KeychainDeletable: KeychainQuery {
+    var deleteQuery: [String: Any] { get }
+}
+
+extension KeychainGettable {
+    var getQuery: [String: Any] {
+        var result = self.query
+        result[kSecMatchLimit as String] = kSecMatchLimitOne
+        result[kSecReturnData as String] = kCFBooleanTrue
+        return result
+    }
+
+    func performGet(prompt: String?, authenticationContext: KeychainAuthenticationContext?) throws -> Keychain.Item {
+        var cfQuery = self.getQuery
+        if let prompt = prompt {
+            cfQuery[kSecUseOperationPrompt as String] = prompt
+        }
+        
+        if let authenticationContext = authenticationContext {
+            cfQuery[kSecUseAuthenticationContext as String] = authenticationContext
+        }
+
+        var ref: AnyObject?
+        
+        let status = Keychain.implementation.copyItemMatching(cfQuery as CFDictionary, &ref)
+        guard status == noErr else {
+            if status == errSecItemNotFound {
+                throw KeychainError.notFound
+            } else {
+                throw KeychainError.cannotGet(code: status)
+            }
+        }
+        
+        guard let result = ref as? Dictionary<String, Any> else {
+            throw KeychainError.invalidFormat
+        }
+        
+        return try Keychain.Item(result)
+    }
+}
+
+extension KeychainListable {
+    var listQuery: [String: Any] {
+        var result = self.query
+        result[kSecMatchLimit as String] = kSecMatchLimitAll
+        return result
+    }
+
+    func performList() throws -> [Keychain.Search.Result] {
+        let cfQuery = self.listQuery as CFDictionary
+        var ref: CFTypeRef?
+        let status = Keychain.implementation.copyItemMatching(cfQuery, &ref)
+        
+        guard status != errSecItemNotFound else {
+            return []
+        }
+        
+        guard status == errSecSuccess else {
+            throw KeychainError.cannotList(code: status)
+        }
+        
+        guard let items = ref as? Array<Dictionary<String, Any>> else {
+            throw KeychainError.invalidFormat
+        }
+
+        return try items.map { try Keychain.Search.Result($0) }
+    }
+}
+
+extension KeychainUpdatable {
+    var updateQuery: [String: Any] {
+        query.filter { (key: String, value: Any) in
+            let keyAttr = key as CFString
+            switch keyAttr {
+            case kSecClass: fallthrough
+            case kSecAttrService: fallthrough
+            case kSecAttrServer: fallthrough
+            case kSecAttrAccessGroup: fallthrough
+            case kSecAttrAccount:
+                return true
+            default:
+                return false
+            }
+        }
+    }
+
+    func performUpdate(_ item: Keychain.Item, authenticationContext: KeychainAuthenticationContext?) throws {
+        let updateSearchQuery = self.updateQuery
+
+        var saveQuery = item.query
+        saveQuery.removeValue(forKey: kSecClass as String)
```

Comment:
BTW, why we are doing remove if our underlying call is `SecItemUpdate`? User may loose previous data if `SecItemUpdate` fails